### ISSUE TITLE
Crash at startup on InkPad 3 device

### ISSUE THREAD

**Issue by dekani** [koreader/koreader#5861]
* KOReader version: v2020.01-42-g3ca4657_2020-02-12
* Device: Tea InkPad 3 (based on PocketBook firmware)
* Firmware: T740.5.17.1680 (Build date: 20180502_143909)

#### Issue
KOReader generates a segmentation fault at startup on my Inkpad 3 device with latest nighlty builds.
If libcurl.so.4 library is deleted or renamed, KOReader starts successfully.

#### Steps to reproduce
Just start KOReader.

##### Attached files :
- crash.log
- log.txt.gz: strace log file. (command used to generate it : strace -f -o log.txt koreader.app)
[crash.log](https://github.com/koreader/koreader/files/4200894/crash.log)
[log.txt.gz](https://github.com/koreader/koreader/files/4200895/log.txt.gz)

**Comment by NiLuJe** [comment]
Okay, something's definitely loading libcurl, which shouldn't happen.

It's also trying to load OpenSSL 1.0 (and ends up loading the system's copy), while we ship 1.1.x, so that's not great...

And that something is prooooobably inkview itself, which is... unfortunate.

Add to that the fact that it's C++, so it also loads the local libstdc++, and that's a lot of reasons for stuff to go boom....

@dekani: Which FW version is that? Are they handled by PocketBook or by whoever's handling Tea (I can never remember who's behind it in France, Decitre? Chapitre? Not that it matters ;D).

**Comment by NiLuJe** [comment]
I can probably attempt to avoid C++ ABI issues, so the STL issue should hopefully be harmless.

Nothing can be done about the OpenSSL thing, AFAIK, which is extremely unfortunate, because it might suddenly start to blow up in fun and interesting ways, too.

As for libcurl, the only stupid thing I can think of is to rename ours (and modify zsync2 accordingly), so that InkView finds and loads the one it expects.

----

Other than that, using a local copy of inkview with mangled rpaths so that it prefers system paths?

Or restoring the standard LD_LIBRARY_PATH when we're loading inkview from Lua?

**Comment by NiLuJe** [comment]
@Frenzie: It's an [ffi.load](https://github.com/koreader/koreader/blob/3e00e686b6fe76709fd6f0ac1696eebbac1e9fb5/frontend/device/pocketbook/device.lua#L4), at least, so would my stupid idea of clearing LD_LIBRARY_PATH if the target lib is inkview be doable/sound sensible?

**Comment by NiLuJe** [comment]
Note that InkView also loads libjpeg (ours, so, might also blow up one day), libtiff (system's), libz (ours, again, strong kablooey potential), libpng (system's, because older sover; again, potentially terrible results), freetype (ours); ICU (system's, because we mercifully don't ship it).

TL;DR: That's a massive Jenga tower of fail waiting to happen.

**Comment by NiLuJe** [comment]
So, yeah, the solution that has the least potential to rock the boat would *probably* be to rename libcurl on PB.

Still worth testing the LD_LIBRATY_PATH/inkview thing, though ;).

**Comment by NiLuJe** [comment]
Hmm, or copying inkview in our libs folder, and neutering the DT_NEEDED entries in that copy (somewhat fragile, but might work).

**Comment by NiLuJe** [comment]
InkView appears to be doing a lot of fun and completely unneeded crap behind our back, too, which might impede my previous idea. Yay -_-".

**Comment by Frenzie** [comment]
@NiLuJe Not entirely sure, but we perform some tricks with ffi.load itself in any case:

https://github.com/koreader/koreader/blob/3e00e686b6fe76709fd6f0ac1696eebbac1e9fb5/setupkoenv.lua#L10-L32

**Comment by dekani** [comment]
About TEA firmware version, it is a version customized to include Cultura store. (buyed there too few years ago)
I have updated first post with more firmware details.
Firmware is no more updated, not really supported now and it isn't synchronized with Official PocketBook firmware versions.
I haven't tried to update to a PocketBook firmware version for many reasons. (store, official TEA firmware isn't available for download and only available by OTA upgrade, users have tried to installed it and have many display issues and can't go back to native TEA firmware)
This firmware also works fine for me. (I don't encounter most of open issues for this device - no slowdown but most of the time, only KOReader is running - only wifi issue but it isn't an annoying issue as starting a native web browser process in background before launching KOReader fix it - software upgrades, OPDS Catalog browsing/download, ... work - as described in Mobileread forum. I will see if I can do more tests on this issue too to provide more information)

**Comment by pazos** [comment]
> So, yeah, the solution that has the least potential to rock the boat would probably be to rename libcurl on PB.

That sounds fine. My suggestion, removing libcurl+zsync2 and foce PB users to update manually, not so fine 🌵

**Comment by NiLuJe** [comment]
Do you know which version it's based on? Or what is the current version
called?

**Comment by NiLuJe** [comment]
@dekani:

Okay, so, one test for each of my stupid ideas ^^.

* First, the patched inkview. Start with a stock install (i.e., with libcurl), extract in the koreader folder.
[PB_5861_Test_1.zip](https://github.com/koreader/koreader/files/4205895/PB_5861_Test_1.zip)

This is based on InkView from FW 5.17, hopefully that won't be an issue. Ideally, this should be based on *your* InkView, but this was faster as a PoC.
It might blow up because of the patch, or because of the version mismatch. I could use more details about your FW version (or a  copy of your own /usr/lib/libinkview.so) ;).

I don't think I'll need any logs from that one.

[PB_5861_Test_1.zip](https://github.com/koreader/koreader/files/4205919/PB_5861_Test_1.zip)



* Second, the ffi.load hack. Start with a stock install (i.e., with libcurl, without Test 1).

I'll need the strace log to confirm how it behaves ;). If it boots, try opening an ePub, see if that doesn't crash.

[PB_5861_Test_2.zip](https://github.com/koreader/koreader/files/4205920/PB_5861_Test_2.zip)

* Third, the renamed libcurl. Start with a stock install (no previous tests applied), delete the original libs/libcurl.so.4

I don't think I'll need any logs from that one, but do try to open an ePub if it boots (which it should) ;).

**Comment by NiLuJe** [comment]
For posterity's sake, test 2's patch:

```diff
--- setupkoenv.lua      2019-03-29 00:52:31.802131788 +0100
+++ /var/tmp/niluje/Test_2/setupkoenv.lua       2020-02-14 18:45:15.004769570 +0100
@@ -16,9 +16,22 @@ end
 local ffi_load = ffi.load
 -- patch ffi.load for thirdparty luajit libraries
 ffi.load = function(lib)
-    io.write("ffi.load: ", lib, "\n")
-    local loaded, re = pcall(ffi_load, lib)
-    if loaded then return re end
+    -- PB hack (#5861)
+    if lib == "inkview" then
+        -- Save the current LD_LIBRARY_PATH
+        local ld_lib_path = os.getenv("LD_LIBRARY_PATH")
+        -- Clear it
+        C.unsetenv("LD_LIBRARY_PATH")
+        io.write("ffi.load: ", lib, " (system's searchpath)\n")
+        local loaded, re = pcall(ffi_load, lib)
+        -- Restore it
+        C.setenv("LD_LIBRARY_PATH", ld_lib_path, 1)
+        if loaded then return re end
+    else
+        io.write("ffi.load: ", lib, "\n")
+        local loaded, re = pcall(ffi_load, lib)
+        if loaded then return re end
+    end
 
     local lib_path = package.searchpath(lib, "./lib?.so;./libs/lib?.so;./libs/lib?.so.1")
```

**Comment by Frenzie** [comment]
@dekani Could you also try this build? It's based on https://github.com/koreader/koreader/pull/5860, https://github.com/koreader/koreader-base/pull/1048, https://github.com/koreader/koxtoolchain/pull/22 and https://github.com/koreader/virdevenv/pull/49

https://mega.nz/#F!HV4jFaKJ

**Comment by dekani** [comment]
@NiLuJe 
Is FW part of firmware version added in first post? 5.17.1680 ?
T740.5.17.1680 (Build date: 20180502_143909)

Test 1:
From crash.log file:
Inconsistency detected by ld.so: dl-version.c: 230: _dl_check_map_versions: Assertion `needed != ((void *)0)' failed!
No strace added as ld fails.
I have attached my own libinkview.so library as attached file.

Test 2:
Same segmentation fault and strace seems to be similar to previous one. (added as attached file)

Test 3:
KOReader starts and epub file is correctly open.

@Frenzie  I will download and test it.

[libinkview.so.gz](https://github.com/koreader/koreader/files/4206875/libinkview.so.gz)
[log-test2.txt.gz](https://github.com/koreader/koreader/files/4206876/log-test2.txt.gz)

**Comment by pazos** [comment]
@dekani: the definitive test would be the package @Frenzie linked with @NiLuJe's Test3. If that works it would be cool

**Comment by pazos** [comment]
Also `libcurl_ko.so.4` is an excellent name for a shared library :sweat_smile:

**Comment by dekani** [comment]
Just tested it...
Same segmentation fault in crash.log file. ([*] Version: v2020.01-36-gf8a34ce_2020-02-12)
I have also generated a strace file if needed but content is similar to previous ones. (crash after ssl initializations)

[log-build.txt.gz](https://github.com/koreader/koreader/files/4206944/log-build.txt.gz)

**Comment by dekani** [comment]
I have done all my previous tests based on this version : koreader-pocketbook-v2020.01-31-ge49198c_2020-02-11.zip as stock version.

**Comment by dekani** [comment]
When is used libcurl library?
When libcurl library is removed or renamed, software updates and OPDS catalog browsing/ file download are still working.

**Comment by pazos** [comment]
@dekani:

If I understood correctly:

1. You downloaded and installed stock version v2020.01-36-gf8a34ce_2020-02-12 from the mega link
2. You deleted the original files zsync2 and libs/libcurl.so.4
3. You placed new objects contained in Test3 zipfile (zsync and libs/libcurl_ko.so.4).
4. With that setup and no other changes you manage to produce a SEGFAULT

If I didn't understand correctly please try these 4 steps.

**Comment by pazos** [comment]
libcurl is a dependency to zsync2. It is used for Over The Air Updates.

**Comment by dekani** [comment]
@pazos I have done 4 steps with koreader-pocketbook-v2020.01-31-ge49198c_2020-02-11.zip as stock version.
Sorry, I have tested v2020.01-36-gf8a34ce_2020-02-12 build without Test3.  I'm testing it...

**Comment by dekani** [comment]
@pazos Yes, it works with v2020.01-36-gf8a34ce_2020-02-12 and Test3. (all your 4 steps - OTA update to v2020.02 works)

But what is strange is OTA update to 2020.02 still works too with this build version if I only remove libcurl library. (I haven't seen any other libcurl library in system library folders, I will check with strace to confirm)

**Comment by pazos** [comment]
@dekani: thanks for the tests.

Is not strange. You used zync2 binary to upgrade/downgrade to other version. The new version overrides zsync but not `libcurl_ko.so.4`, so after the update you have both libraries in libs/ but the "new" zsync2 binary will try to use the unvendored one, so you won't be able to do a new OTA update from that firmware unless you copy Test3 files again (and remove unvendored libcurl to avoid the segfault).

**Comment by dekani** [comment]
Yes.
But only for information, I have done a test with an old zip package (clean install) and it works too without libcurl library file. (koreader-pocketbook-v2020.01-28-ga2ea88d_2020-02-10.zip, it already contains zsync2)
 
I can't add full strace (too big), but it finally found libcurl library in /usr/lib folder too (I haven't search correctly before) when library is removed from koreader libs folder and I can confirm OTA update works too.

1665  open("/mnt/ext1/applications/koreader/libs/libcurl.so.4", O_RDONLY) = -1 ENOENT (No such file or directory)
1665  open("libcurl.so.4", O_RDONLY)    = -1 ENOENT (No such file or directory)
1665  open("/lib/libcurl.so.4", O_RDONLY) = -1 ENOENT (No such file or directory)
1665  open("/usr/lib/libcurl.so.4", O_RDONLY) = 3
...
1695  send(4, "GET /koreader-pocketbook-v2020.0"..., 117, MSG_NOSIGNAL) = 117

bash-4.1$ ls /usr/lib/libcurl*
libcurl.so
libcurl.so.4
libcurl.so.4.5.0

Renaming it is clearly a better choice to avoid using system one...

**Comment by NiLuJe** [comment]
Gah, it seems that patchelf still breaks binaries in a few cases, I was afraid of that as I've had trouble with it in the past, but it seemed like they'd fixed it. That said, inkview is built in a very unorthodox manner, which probably doesn't help.

That was for test 1 ;).

----

Test 2 didn't pan out, which is strange, do you happen to have the matching crash.log so I can double-check something?

----

Test 3 is trivial in practice, so, not terribly surprised it works.

----

It's entirely possible that our own zsync2 might behave properly with the system's libcurl. But it's not necessarily a given. And it's fragile, at best. Hence making sure it uses our own with the renaming trick ;).

**Comment by NiLuJe** [comment]
Note to self: don't strip DT_NEEDED entries for libraries with explicit symbol versioning mentioned in the `.gnu.version_r` ssection (c.f., https://sourceware.org/bugzilla/show_bug.cgi?id=16452).

Here, that's libpng12 & libjpeg, which are potentially less problematic (libpng12 is a separate sover w/ versioned symbols; libjpeg pulls our own, hopefully the ABI/API is stable, because JPEG, and that's typically the kind of crap a lot of projects w/ binary releases must have been pulling off dfor quite a while ;)).

**Comment by NiLuJe** [comment]
Okay, armed with that new knowledge (and quite a few horror stories from the patchelf & LIEF bug trackers :D), here's a variant of Test 1 that prooobably doesn't horribly blow up in the dynamic loader ;).

[PB_5861_Test_1b.zip](https://github.com/koreader/koreader/files/4207606/PB_5861_Test_1b.zip)

----

Let's say I'll need both the crash.log and a strace, because apparently I forgot to say that for Test 2 earlier and now I'm sad that I don't have it ;p.

**Comment by NiLuJe** [comment]
Oh, I think I know why Test 2 did not pan out: `base/input/input-pocketbook.h` is *also* `dlopen`ing inkview :s.

That would require patching a [`require`](https://github.com/koreader/koreader-base/blob/019aa33ad512efc49209a4381308c9b50e9892f7/ffi/input.lua#L10) instead of an `ffi.load`, so,eeeh, let's forget about it.

**Comment by NiLuJe** [comment]
@dekani: Okay, here's Test 2b ;). Same as the previous iteration, start from a stock setup. Both the crash.log & strace log will be useful ;).

[PB_5861_Test_2b.zip](https://github.com/koreader/koreader/files/4207691/PB_5861_Test_2b.zip)

**Comment by dekani** [comment]
@NiLuJe Each patches have been applied after a clean koreader-pocketbook-v2020.02.zip installation.

Here are the results:
[crash_test1b.log.gz](https://github.com/koreader/koreader/files/4207928/crash_test1b.log.gz)
[log_test1b.txt.gz](https://github.com/koreader/koreader/files/4207929/log_test1b.txt.gz)
[crash_test2b.log.gz](https://github.com/koreader/koreader/files/4207930/crash_test2b.log.gz)
[log_test2b.txt.gz](https://github.com/koreader/koreader/files/4207931/log_test2b.txt.gz)

**Comment by NiLuJe** [comment]
Okay, so Test 1b fails possibly because there are some global *variables* that refer to curl (and/or it really sets up a curl handle at init :s).

And Test 2b really didn't take for... some reason. Not sure why. luajit itself doesn't have an rpath set. Eh.

----

So, that leaves option 3 :/.

I'll see how to actually mangle that in our buildsystem them. (Yaaaaaaay -_-").

**Comment by NiLuJe** [comment]
@dekani: Okay! [this build](http://files.ak-team.com/niluje/mrpub/Other/koreader-pocketbook-v2020.01-36-gf8a34ce5_2020-02-12.zip) *hopefully* implements Option 3 properly, so a fresh install of it *should* Just Work (TM) ;).

**Comment by dekani** [comment]
@NiLuJe Yes, perfect. It works fine. (OTA update to 2020.02 works too)
I have developped C/C++ softwares for Unix OS (Linux and AIX) and do docker containers, it helps and I do french translations for projects I appreciate. :)

Thanks, everybody.

**Comment by Frenzie** [comment]
Please drop by our French translation. ;-)

**Comment by dekani** [comment]
Yes, done since few months. :) Migrated to weblate too.

**Comment by Frenzie** [comment]
Thanks! I've noticed the French translation seems to be up to date these days. :-D

**Comment by mikkelnl** [comment]
Hi guys, sorry but I'm not sure I understand your updates (I'm not that tech savvy), is it OK to OTA update my Pocketbook Touch HD 3 to the new 2020.02 version? :) Thanks :)

**Comment by NiLuJe** [comment]
The fix hasn't made it to a release or nightly yet.
